dig.go

```diff
@@ -132,36 +147,36 @@ func (c *Container) provide(ctor interface{}, ctype reflect.Type) error {
 }
 
 // Get the return types of a constructor with all the dig.Out returns get expanded.
-func (c *Container) getReturnTypes(
+func (c *Container) getReturnKeys(
 	ctor interface{},
 	ctype reflect.Type,
-) (map[reflect.Type]struct{}, error) {
+) (map[key]struct{}, error) {
 	// Could pre-compute the size but it's tricky as counter is different
 	// when dig.Out objects are mixed in
-	returnTypes := make(map[reflect.Type]struct{})
+	returnTypes := make(map[key]struct{})
 
 	// Check each return object
 	for i := 0; i < ctype.NumOut(); i++ {
 		outt := ctype.Out(i)
 
-		err := traverseOutTypes(outt, func(rt reflect.Type) error {
-			if rt == _errType {
+		err := traverseOutTypes(key{t: outt}, func(k key) error {
+			if k.t == _errType {
 				// Don't register errors into the container.
 				return nil
 			}
 
 			// Tons of error checking
-			if isInObject(rt) {
+			if isInObject(k.t) {
 				return errors.New("can't provide parameter objects")
 			}
-			if _, ok := returnTypes[rt]; ok {
-				return fmt.Errorf("returns multiple %v", rt)
+			if _, ok := returnTypes[k]; ok {
+				return fmt.Errorf("returns multiple %v", k.t)
 			}
-			if _, ok := c.nodes[rt]; ok {
-				return fmt.Errorf("provides type %v, which is already in the container", rt)
+			if _, ok := c.nodes[k]; ok {
+				return fmt.Errorf("provides type %v, which is already in the container", k.t)
```

Comment:
That's actually a good point - error messaging.

But I think we can do it even better by just including the `key` to identify which object fails (that already handles for printing out the name as well, if it's present)